Don't switch back to original branch at the end of `brew update`

ilovezfs opened the pull request:
Issue #48812

ilovezfs commented:
@mikemcquaid Do you also want this behavior restricted to developers, or should it be left as it is?
https://github.com/Homebrew/homebrew/blob/3ef1fb33a3b356d7f61e9b51ac830e969c8cc4aa/Library/Homebrew/cmd/update.rb#L306-L318

xu-cheng commented on Library/Homebrew/cmd/update.rb:
```diff
@@ -277,8 +277,7 @@ def pull!(options = {})
     @current_revision = read_current_revision
-
-    if @initial_branch != @upstream_branch && !@initial_branch.empty?
+    if ARGV.homebrew_developer? && @initial_branch != @upstream_branch && !@initial_branch.empty?
```
Could we use a different flag or ENV variable for this? It is counter-intiative to have `HOMEBREW_DEVELOPER` uses a complete different update strategy.

mikemcquaid replied:
Perhaps we should just remove this logic altogether for everyone. I don't think changing back to a branch is too much work.

xu-cheng replied:
I think `pop_stash` is still very useful. Could we have a logical similar like this?
```
if pop stash will create conflict
  show pop stash message
else
  pop stash automatically for user
end
```

mikemcquaid replied:
I'd definitely like to avoid popping the stash for end-users. Note that it only prints the message unless you were already on a non-default branch; I think we can just print the message in that case too.

xu-cheng replied:
:cool:

mikemcquaid commented:
> @mikemcquaid Do you also want this behavior restricted to developers, or should it be left as it is?
> https://github.com/Homebrew/homebrew/blob/3ef1fb33a3b356d7f61e9b51ac830e969c8cc4aa/Library/Homebrew/cmd/update.rb#L306-L318

That behaviour should be left as-is. Once we figure out how we want to approach this we should maybe make that consistent.

ilovezfs commented:
@xu-cheng @mikemcquaid Should branch-switch-logic be separated from pop-stash-logic entirely?

A) back to initial branch, pop
B) back to initial branch, pop if no conflict
C) back to initial branch, print message
D) stay on master, pop
E) stay on master, pop if no conflict
F) stay on master, print message

Currently this implements
developer ? A : F

@mikemcquaid Are you saying you'd prefer
developer ? C : F
or are you saying something else?

@xu-cheng Are you saying you'd prefer
new-option ? B : F
or are you saying something else?

Or should we do
new-option' && new-option'' -> B
new-option' && ~new-option'' -> C
~new-option' && new-option'' -> E
~new-option' && ~new-option'' -> F [default]

mikemcquaid commented:
@ilovezfs I think it'd be good to remove all "switch back from `master`" logic.

ilovezfs commented:
@mikemcquaid OK, so we have

X) pop
Y) pop if no conflict
Z) print message

So you're saying you'd prefer
developer ? X : Z
or
developer ? Y : Z
or
new-option ? X : Z
or
new-option ? Y : Z
or
new-option=X ? X : Z
new-option=Y ? Y: Z
new-option=Z ? Z : Z
or
something else?

mikemcquaid commented:
I prefer always Z and no option.

ilovezfs commented:
@mikemcquaid Me like.

ilovezfs commented:
@mikemcquaid @xu-cheng OK, PR is refreshed.

mikemcquaid commented:
Looks good @ilovezfs! If you want additional :sparkles:: could you make the same change in `update-bash.sh`?

ilovezfs commented:
@mikemcquaid Sure, I'll take a look at update-bash.sh

I'm a little concerned that the messages printed need to mention branch switches and also include them in the pop stash directions. Here's what's going to happen in support tickets now:
```
bash-4.3$ cd `brew --prefix`
bash-4.3$ git status
On branch mybranch
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   Library/hello.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Library/nottracked2.txt

no changes added to commit (use "git add" and/or "git commit -a")
bash-4.3$ git log -n1
commit c2897c91704df5a865c34fad6de6df0e29ffb3b8
Author: ilovezfs <ilovezfs@icloud.com>
Date:   Tue Feb 9 06:29:38 2016 -0800

    add hello.txt
bash-4.3$ cat Library/hello.txt 
woof
meow
bash-4.3$ git diff
diff --git a/Library/hello.txt b/Library/hello.txt
index cfbc74a..3ca8411 100644
--- a/Library/hello.txt
+++ b/Library/hello.txt
@@ -1 +1,2 @@
 woof
+meow
bash-4.3$ brew update
To restore the stashed changes to /usr/local run:
  `cd /usr/local && git stash pop`
Already up-to-date.
bash-4.3$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
bash-4.3$ #I'm just following orders
bash-4.3$ cd /usr/local && git stash pop
CONFLICT (modify/delete): Library/hello.txt deleted in Updated upstream and modified in Stashed changes. Version Stashed changes of Library/hello.txt left in tree.
bash-4.3$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add/rm <file>..." as appropriate to mark resolution)

	deleted by us:   Library/hello.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Library/nottracked2.txt

no changes added to commit (use "git add" and/or "git commit -a")
bash-4.3$ #oh yeah my branch
bash-4.3$ git checkout mybranch
Library/hello.txt: needs merge
error: you need to resolve your current index first
```

ilovezfs commented:
@mikemcquaid PR refreshed to include update-bash.sh as well.

mikemcquaid commented:
Looks great, thanks!

UniqMartin commented:
I already see myself getting confused all the time, because I'm almost never on the `master` branch when I run the update, but oh well. I guess I'll adapt. :+1:

ilovezfs commented:
@UniqMartin Maybe
```
brewup(){(branch=$(basename $(/usr/bin/git -C "$(brew --prefix)" symbolic-ref HEAD)); brew update ; /usr/bin/git -C "$(brew --prefix)" checkout $branch)}
```

UniqMartin commented:
@ilovezfs Thanks for the suggestion! But don't worry about me; I'll either adapt to the new behavior or else know my way around the command line well enough to come up with something like you suggested. And thanks for creating this PR in the first place!

mikemcquaid commented:
Thanks again @ilovezfs!

ilovezfs commented:
You're welcome! Alea iacta est.

mikemcquaid commented:
To save anyone else looking it up:

> Alea iacta est ("The die is cast") is a Latin phrase attributed by Suetonius (as iacta alea est [ˈjakta ˈaːlea est]) to Julius Caesar on January 10, 49 BC as he led his army across the Rubicon river in Northern Italy.

https://en.wikipedia.org/wiki/Alea_iacta_est
